hamilton/driver.py

```diff
@@ -638,20 +638,80 @@ def raw_execute(
         return out
 
     @capture_function_usage
-    def list_available_variables(self) -> List[Variable]:
+    def list_available_variables(
+        self, *, tag_filter: Dict[str, Union[Optional[str], List[str]]] = None
+    ) -> List[Variable]:
         """Returns available variables, i.e. outputs.
 
-        These variables corresond 1:1 with nodes in the DAG, and contain the following information:
-        1. name: the name of the node
-        2. tags: the tags associated with this node
-        3. type: The type of data this node returns
-        4. is_external_input: Whether this node represents an external input (required from outside),
-        or not (has a function specifying its behavior).
+        These variables correspond 1:1 with nodes in the DAG, and contain the following information:
+
+            1. name: the name of the node
+            2. tags: the tags associated with this node
+            3. type: The type of data this node returns
+            4. is_external_input: Whether this node represents an external input (required from outside),
+            or not (has a function specifying its behavior).
 
+        .. code-block:: python
 
+            # gets all
+            dr.list_available_variables()
+            # gets exact matching
+            dr.list_available_variables({"TAG_NAME": "TAG_VALUE"})
```

Comment:
This API is confusing me quite a bit TBH. `TAG_VALUE` is a list means that they *all* have to match? BUt if its an individual it means they have to match any?

We should be opinionated and simple here. If not, we should leave querying up to the user... I was thinking:
1. Only query for individual tag_values in dict. Dicts imply `and`, and an individual one means its one of the decorated values
2. We leave open the option for more kw params later

```python
dr.select_variables(tag_match={"foo" : "bar", "baz": "qux"})
```

This would match:
- `@tag(foo="bar")`
- `@tag(baz="qux")`
- `@tag(foo=["bar", "qux"])

But not:
- `@tag(foo="not_bar")`
- `@tag(foo=["not_baz", "not_bar"])`
- `@tag(foo="bar", baz="not_qux")`
- `@tag(foo=["bar", "qux"], baz=["bar", "not_qux"])`

This is being opinionated that matching means any do exist, anything else is up to the user to specify/query independently.

Reply:
The issue is how do you interpret a list of tag values?

E.g. Given:

```python
@tag(business_lines=["CA", "TX"])
def my_func() -> ...:
   pass

```
```python
# e.g. I want to get all nodes with a particular business line (it could be variable in length)
# should this be an exact match if someone passes a single item list? 
dr.list_available_variables(tag_filter=dict(business_lines=["CA"]))

# to me the above is equivalent to this -- because a user could also write this:
dr.list_available_variables(tag_filter=dict(business_lines="CA"))

# Why?
# Well, why would someone put a tag on a function with only a single value in it?
@tag(business_lines=["CA"]) # < -- this wouldn't make sense to me.
def my_func() -> ...:
   pass

# e.g. I want to get all nodes that correspond to Cali AND Texas AND Florida:
# what about this? I think this is an exact match requirement (i.e. AND), not an OR.
dr.list_available_variables(tag_filter=dict(business_lines=["CA", "TX", "FL"]))

# e.g. I want to get all nodes that hit Cali OR Texas OR Florida
# If you want an OR you would have to do multiple calls and union the results.
(
 dr.list_available_variables(tag_filter=dict(business_lines="CA"))  | 
 dr.list_available_variables(tag_filter=dict(business_lines="TX"))  | 
 dr.list_available_variables(tag_filter=dict(business_lines="FL"))  | 
)
```

Reply:
@kzk2000 any thoughts on lists of values? It would be simpler to not allow it.

Reply:
@skrawcz Whatever feels more intuitive is all right by me.

Say
```
@tag( business_value=["CA", "US"] )`
def combo_node():
    ...

@tag( business_value=["US"] )`
def only_us_node():
    ...

@tag( business_value=["CA"] )`
def only_ca_node():
    ...
```


That is, I would treat the last case as OR operator, if any one of the list entries matches, return that node.
I'm totally fine with making this more explicit as you suggested here.
```
(
 dr.list_available_variables(tag_filter=dict(business_lines="CA"))  | 
 dr.list_available_variables(tag_filter=dict(business_lines="TX"))  | 
 dr.list_available_variables(tag_filter=dict(business_lines="FL"))  | 
) # you'd have to make them sets for this to work but you get the picture
```

Bottom line as long as `dr.list_available_variables()``` returns all nodes, I can easily run any list comprehension filter like this myself.

[ x for x in  dr.list_available_variables() if x.tag.get('business_line", None) in ['US', 'CA']]

or use sets as you suggested

Reply:
Okay I think I might switch to the OR interpretation then.

Reply:
Will try to have something by Friday

Reply:
@elijahbenizzy @kzk2000 

For this case:
```python
dr.list_available_variables(tag_filter=dict(business_lines=["CA"], some_other_tag="foo" ))
```
I'm interpreting it as an AND query, while the values in the list are OR. Thoughts?

Reply:
So yeah, offline we've synced up and decided that tags are *or* queries. This leaves the door open with:
```python
tag_filter=dict(foo=or_("bar", "baz"))
tag_filter=dict(foo=and_("bar", "baz"))
```

or something like this

Reply:
To confirm:


> dr.list_available_variables(tag_filter=dict(business_lines=["CA", "US"], some_other_tag="foo" ))

this to me reads:
    business_lines=CA  OR business_lines=US 
AND 
    some_other_tag=foo